### PR TITLE
Rebase then Incremental Unit Tests

### PR DESCRIPTION
**Pull request by pfaraone**
## Changes
- Unit testing for rebasing follow by incremental compaction
- Iterable CompactorTestCase class and subclasses defined so test cases parameters can be passed via constructor kwargs.
- `setup_local_deltacat_storage_conn` fixture now cleans up the table as part of its teardown procedure (after `yield`) instead of using an explicit teardown fixture

### CODE REVIEW

File: deltacat/tests/compute/compact_partition_test_cases.py [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+        create_table_strategy=create_src_w_deltas_destination_strategy,
+    ),
+    "9-incremental-decimal-pk-multi-dup": IncrementalCompactionTestCase(
+        primary_keys_param={"pk_col_1"},
+        sort_keys_param=[
+            {
+                "key_name": "sk_col_1",
+            },
+        ],
+        partition_keys_param=[{"key_name": "region_id", "key_type": "int"}],
```
Comment (rkenmi) [inline deltacat/tests/compute/compact_partition_test_cases.py:312]: Should we also include partition-less tables?

File: deltacat/tests/compute/compact_partition_test_cases.py [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+            final_cases[f"{case_name}_v{version + 1}"] = [
+                *case_value,
+                compact_partition_func,
+            ]
+
+    return final_cases
+
+
+@dataclass(frozen=True)
+class CompactorTestCase:
```
Comment (rkenmi) [inline deltacat/tests/compute/compact_partition_test_cases.py:43]: Could we have docstrings that describe each of these test specific parameters?

File: deltacat/tests/compute/compact_partition_test_cases.py [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+    partition_keys_param: List[Dict[str, str]]
+    partition_values_param: str
+    column_names_param: List[str]
+    input_deltas_arrow_arrays_param: List[pa.Array]
+    input_deltas_delta_type: str
+    expected_terminal_compact_partition_result: pa.Table
+    create_placement_group_param: bool
+    records_per_compacted_file_param: int
+    hash_bucket_count_param: int
+    validation_callback_func: Callable
```
Comment (rkenmi) [inline deltacat/tests/compute/compact_partition_test_cases.py:55]: What is `validation_callback_func` used for? Is it used in any test?

File: deltacat/tests/compute/test_compact_partition_rebase_then_incremental.py [hunk 4]
```diff
@@ -69,239 +76,183 @@ def setup_compaction_artifacts_s3_bucket(setup_s3_resource: ServiceResource):
...
     )
     from deltacat.utils.placement import (
         PlacementGroupManager,
     )
-    from deltacat.compute.compactor import (
-        RoundCompletionInfo,
-    )
 
     ds_mock_kwargs = setup_local_deltacat_storage_conn
+    ray.shutdown()
```
Comment (rkenmi) [inline deltacat/tests/compute/test_compact_partition_rebase_then_incremental.py:193]: Why does ray need to be shut down and switched to local? Can this be handled during test setup instead?

File: deltacat/tests/compute/compact_partition_test_cases.py [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+
+    # makes CompactorTestCase iterable which is required to build the list of pytest.param values to pass to pytest.mark.parametrize
+    def __iter__(self):
+        return (getattr(self, field.name) for field in fields(self))
+
+
+@dataclass(frozen=True)
+class IncrementalCompactionTestCase(CompactorTestCase):
+    pass
+    # create_table_strategy = lambda *args, **kwargs: create_table_for_incremental_case_strategy(**kwargs)
```
Comment (rkenmi) [inline deltacat/tests/compute/compact_partition_test_cases.py:67]: Minor: Unused comment

File: deltacat/tests/compute/compact_partition_test_cases.py [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+        ),
+        validation_callback_func=None,
+        validation_callback_func_kwargs=None,
+        create_placement_group_param=True,
+        records_per_compacted_file_param=DEFAULT_MAX_RECORDS_PER_FILE,
+        hash_bucket_count_param=None,
+        create_table_strategy=create_src_w_deltas_destination_rebase_w_deltas_strategy,
+    ),
+}
+
```
Comment (raghumdani) [inline deltacat/tests/compute/compact_partition_test_cases.py:442]: Can you also add a test case with multiple primary keys?

File: deltacat/tests/compute/compact_partition_test_cases.py [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+                pa.array(["foo"] * 10),
+                pa.array([i / 10 for i in range(40, 50)]),
+            ],
+            names=["pk_col_1", "sk_col_1", "sk_col_2", "col_1"],
+        ),
+        validation_callback_func=None,
+        validation_callback_func_kwargs=None,
+        create_placement_group_param=True,
+        records_per_compacted_file_param=DEFAULT_MAX_RECORDS_PER_FILE,
+        hash_bucket_count_param=None,
```
Comment (raghumdani) [inline deltacat/tests/compute/compact_partition_test_cases.py:386]: Can you also add a test case with bigger hash bucket count than records? That way we will know the behavior if some hash buckets don't get utilized.

File: deltacat/tests/compute/test_compact_partition_incremental.py [hunk 8]
```diff
@@ -325,24 +230,17 @@ def test_compact_partition_incremental(
...
-    round_completion_info = RoundCompletionInfo(**rcf_file_output)
-    print(f"rcf_file_output: {json.dumps(rcf_file_output, indent=2)}")
-    compacted_delta_locator = round_completion_info.compacted_delta_locator
     tables = ds.download_delta(compacted_delta_locator, **ds_mock_kwargs)
     compacted_table = pa.concat_tables(tables)
     assert compacted_table.equals(
-        expected_result
-    ), f"{compacted_table} does not match {expected_result}"
+        expected_terminal_compact_partition_result
+    ), f"{compacted_table} does not match {expected_terminal_compact_partition_result}"
```
Comment (raghumdani) [inline deltacat/tests/compute/test_compact_partition_incremental.py:240]: combine_chunks() and sort both the tables before comparing equality. The compactor can create multiple files (which is especially true for v2).

File: deltacat/tests/compute/test_compact_partition_rebase_then_incremental.py [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+    )
+    # execute
+    rcf_file_s3_uri = compact_partition_func(compact_partition_params)
+    compacted_delta_locator = get_compacted_delta_locator_from_rcf(
+        setup_s3_resource, rcf_file_s3_uri
+    )
+    tables = ds.download_delta(compacted_delta_locator, **ds_mock_kwargs)
+    compacted_table = pa.concat_tables(tables)
+    assert compacted_table.equals(
+        rebase_expected_compact_partition_result
```
Comment (raghumdani) [inline deltacat/tests/compute/test_compact_partition_rebase_then_incremental.py:274]: Same here regarding combine_chunks and sort.

File: deltacat/tests/compute/test_compact_partition_rebase_then_incremental.py [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+        }
+    )
+    rcf_file_s3_uri = compact_partition_func(compact_partition_params)
+    compacted_delta_locator = get_compacted_delta_locator_from_rcf(
+        setup_s3_resource, rcf_file_s3_uri
+    )
+    tables = ds.download_delta(compacted_delta_locator, **ds_mock_kwargs)
+    compacted_table = pa.concat_tables(tables)
+    assert compacted_table.equals(
+        expected_terminal_compact_partition_result
```
Comment (raghumdani) [inline deltacat/tests/compute/test_compact_partition_rebase_then_incremental.py:346]: same here regarding combine_chunks and sort.

**Review by raghumdani — COMMENTED** [review]
Looks good overall. Please add more tests.

File: deltacat/tests/compute/compact_partition_test_cases.py (hunk 3 above)
Comment (pfaraone) [reply deltacat/tests/compute/compact_partition_test_cases.py:55]: It was added in #188 for testing extensibility. The test case writer to define additional validations beyond just comparing the expected table values and actual table values after compaction. However, it seems like we haven't had a need for it so it may make sense to remove this test input parameter

File: deltacat/tests/compute/test_compact_partition_rebase_then_incremental.py (hunk 4 above)
Comment (pfaraone) [reply deltacat/tests/compute/test_compact_partition_rebase_then_incremental.py]: Removing the `ray.shutdown()` invocation. We can just use ignore_reinit_error=True argument. There seems to be issues with the unit tests (most of the `compact_partition` test cases are failing) in the `CI` workflow if we handle it as a module-level fixture that seems to be idiosyncratic to the GitHub Actions runner but I haven't yet spent time investigating it. I'll look into it

File: deltacat/tests/compute/compact_partition_test_cases.py (hunk 1 above)
Comment (pfaraone) [reply deltacat/tests/compute/compact_partition_test_cases.py:321]: Done

Comment (pfaraone) [reply deltacat/tests/compute/compact_partition_test_cases.py:321]: Added

```
"10-incremental-decimal-pk-partitionless": IncrementalCompactionTestCase(
        primary_keys={"pk_col_1"},
        sort_keys=[SortKey.of(key_name="sk_col_1")],
        partition_keys_param=ZERO_VALUED_PARTITION_KEYS_PARAM,
        partition_values_param=ZERO_VALUED_PARTITION_VALUES_PARAM,
        column_names_param=["pk_col_1", "sk_col_1"],
        input_deltas_arrow_arrays_param=[
            pa.array([0.1] * 4 + [0.2] * 4 + [0.3] * 4 + [0.4] * 4 + [0.5] * 4),
            pa.array([i for i in range(20)]),
        ],
        input_deltas_delta_type=DeltaType.UPSERT,
        expected_terminal_compact_partition_result=pa.Table.from_arrays(
            [
                pa.array([0.1, 0.2, 0.3, 0.4, 0.5]),
                pa.array([3, 7, 11, 15, 19]),
            ],
            names=["pk_col_1", "sk_col_1"],
        ),
        validation_callback_func=None,
        validation_callback_func_kwargs=None,
        create_placement_group_param=False,
        records_per_compacted_file_param=DEFAULT_MAX_RECORDS_PER_FILE,
        hash_bucket_count_param=None,
        create_table_strategy=create_src_w_deltas_destination_strategy,
    ),
```

File: deltacat/tests/compute/compact_partition_test_cases.py (hunk 2 above)
Comment (pfaraone) [reply deltacat/tests/compute/compact_partition_test_cases.py:51]: Partially added

File: deltacat/tests/compute/test_compact_partition_rebase_then_incremental.py (hunk 4 above)
Comment (pfaraone) [reply deltacat/tests/compute/test_compact_partition_rebase_then_incremental.py:198]: Now fixed - problem was with placement group creation interacting with a resource constrained Github Actions runner.  See https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources

File: deltacat/tests/compute/compact_partition_test_cases.py [hunk 11]
```diff
@@ -0,0 +1,450 @@
...
+            [
+                pa.array(["0", "1", "2", "3", "4", "6"]),
+                pa.array([0, 1, 2, 3, 4, 9]),
+                pa.array(["foo", "foo", "foo", "foo", "foo", "foo"]),
+                pa.array([1, 1.1, 1.2, 1.3, 1.4, 1.9]),
+            ],
+            names=["pk_col_1", "sk_col_1", "sk_col_2", "col_1"],
+        ),
+        incremental_deltas_arrow_arrays_param=[
+            pa.array([str(i) for i in range(5)] + ["6", "6", "7", "7", "8"]),
```
[the remark this answers is not in the document]
Comment (pfaraone) [reply deltacat/tests/compute/compact_partition_test_cases.py:445]: Done 

```
"13-rebase-then-incremental-partial-deltas-on-incremental-deltas": RebaseThenIncrementalCompactorTestCase(
        primary_keys={"pk_col_1"},
        sort_keys=[],
        partition_keys_param=[{"key_name": "region_id", "key_type": "int"}],
        partition_values_param=["1"],
        column_names_param=["pk_col_1", "col_1"],
        input_deltas_arrow_arrays_param=[
            pa.array([str(i) for i in range(10)]),
            pa.array([str(i) for i in range(10)]),
        ],
        input_deltas_delta_type=DeltaType.UPSERT,
        rebase_expected_compact_partition_result=pa.Table.from_arrays(
            [
                pa.array([str(i) for i in range(10)]),
                pa.array([str(i) for i in range(10)]),
            ],
            names=["pk_col_1", "col_1"],
        ),
        incremental_deltas_arrow_arrays_param=[
            pa.array(["9"]),
            pa.array(["100.0"]),
        ],
        incremental_deltas_delta_type=DeltaType.UPSERT,
        expected_terminal_compact_partition_result=pa.Table.from_arrays(
            [
                pa.array([str(i) for i in range(10)]),
                pa.array([str(i) for i in range(9)] + ["100.0"]),
            ],
            names=["pk_col_1", "col_1"],
        ),
        validation_callback_func=None,
        validation_callback_func_kwargs=None,
        create_placement_group_param=False,
        records_per_compacted_file_param=DEFAULT_MAX_RECORDS_PER_FILE,
        hash_bucket_count_param=None,
        create_table_strategy=create_src_w_deltas_destination_rebase_w_deltas_strategy,
    ),
```

File: deltacat/tests/compute/compact_partition_test_cases.py [hunk 12]
```diff
@@ -0,0 +1,544 @@
...
+        validation_callback_func=None,
+        validation_callback_func_kwargs=None,
+        create_placement_group_param=False,
+        records_per_compacted_file_param=DEFAULT_MAX_RECORDS_PER_FILE,
+        hash_bucket_count_param=None,
+        create_table_strategy=create_src_w_deltas_destination_rebase_w_deltas_strategy,
+    ),
+}
+
+INCREMENTAL_TEST_CASES = create_tests_cases_for_enabled_compactor_versions(
```
Comment (raghumdani) [inline deltacat/tests/compute/compact_partition_test_cases.py:537]: Can you add empty delta CSV test case? In that case, compactor should write back empty file.

File: deltacat/tests/compute/compact_partition_test_cases.py (hunk 3 above)
Comment (pfaraone) [reply deltacat/tests/compute/compact_partition_test_cases.py]: Removed

File: deltacat/tests/compute/test_compact_partition_rebase_then_incremental.py [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+        source_table_stream.locator,
+        partition_values_param,
+        **ds_mock_kwargs,
+    )
+    # source_table_partition =
+    # destination_table_partition: Partition = ds.get_partition(
+    #     destination_table_stream,
+    #     partition_values_param,
+    #     **ds_mock_kwargs
+    # )
```
Comment (raghumdani) [inline deltacat/tests/compute/test_compact_partition_rebase_then_incremental.py:221]: Avoid commented out code.

File: deltacat/tests/compute/test_compact_partition_rebase_then_incremental.py (hunk 10 above)
Comment (pfaraone) [reply deltacat/tests/compute/test_compact_partition_rebase_then_incremental.py]: Done

File: deltacat/tests/compute/test_compact_partition_rebase_then_incremental.py (hunk 9 above)
Comment (pfaraone) [reply deltacat/tests/compute/test_compact_partition_rebase_then_incremental.py]: Done

File: deltacat/tests/compute/test_compact_partition_incremental.py (hunk 8 above)
Comment (pfaraone) [reply deltacat/tests/compute/test_compact_partition_incremental.py]: Done

File: deltacat/tests/compute/compact_partition_test_cases.py [hunk 14]
```diff
@@ -0,0 +1,450 @@
...
+            names=["pk_col_1", "sk_col_1", "sk_col_2", "col_1"],
+        ),
+        validation_callback_func=None,
+        validation_callback_func_kwargs=None,
+        create_placement_group_param=True,
+        records_per_compacted_file_param=DEFAULT_MAX_RECORDS_PER_FILE,
+        hash_bucket_count_param=None,
+        create_table_strategy=create_src_w_deltas_destination_rebase_w_deltas_strategy,
+    ),
+}
```
[the remark this answers is not in the document]
Comment (pfaraone) [reply deltacat/tests/compute/compact_partition_test_cases.py:606]: Done

File: deltacat/tests/compute/compact_partition_test_cases.py (hunk 7 above)
Comment (pfaraone) [reply deltacat/tests/compute/compact_partition_test_cases.py]: Done

File: deltacat/tests/compute/compact_partition_test_cases.py (hunk 6 above)
Comment (pfaraone) [reply deltacat/tests/compute/compact_partition_test_cases.py:607]: Done

**Comment by raghumdani** [comment]
Looks good mostly. Please add additional test cases we discussed. Thanks for insisting on highest standards.

**Review by raghumdani — APPROVED** [review]
LGTM
